File: Samples/DeviceEnumerationAndPairing/js/js/scenario1_devicepicker.js
//// Copyright (c) Microsoft Corporation. All rights reserved

(function () {
    "use strict";

    var DevEnum = Windows.Devices.Enumeration;
    var devicePicker = null;
    var resultCollection = new WinJS.Binding.List([]);
    var resultsListView;

    var page = WinJS.UI.Pages.define("/html/scenario1_devicepicker.html", {
        ready: function (element, options) {

            // Setup beforeNavigate event
            WinJS.Navigation.addEventListener("beforenavigate", onLeavingPage);

            // Hook up button event handlers
            document.getElementById("pickSingleDeviceButton").addEventListener("click", pickSingleDeviceClick, false);
            document.getElementById("showDevicePickerButton").addEventListener("click", showDevicePickerClick, false);

            // Hook up result list data binding
            resultsListView = element.querySelector("#resultsListView").winControl;
            resultsListView.itemDataSource = resultCollection.dataSource;

            // Manually bind selector options
            DisplayHelpers.devicePickerSelectors.forEach(function each(item) {
                var option = document.createElement("option");
                option.textContent = item.displayName;
                selectorComboBox.appendChild(option);
            });

            // Process any data bindings
            WinJS.UI.processAll();
        }
    });

    function onLeavingPage() {
        WinJS.Navigation.removeEventListener("beforenavigate", onLeavingPage);
        resultCollection.splice(0, resultCollection.length);
    }

    function pickSingleDeviceClick() {
        showDevicePicker(true);
    }
    function showDevicePickerClick() {
        showDevicePicker(false);
    }

    function showDevicePicker(pickSingle) {
        var buttonRect;

        showDevicePickerButton.disabled = true;
        resultCollection.splice(0, resultCollection.length);

        devicePicker = new DevEnum.DevicePicker();

        var selectedItem = DisplayHelpers.devicePickerSelectors.getAt(selectorComboBox.selectedIndex);

        // If the selected device selector is a pre-canned device class, call CreateWatcher with the device class
        // overload.  Otherwise assume the device selector contains an AQS filter selector and use that overload.
        if (null != selectedItem.deviceClassSelector) {
            devicePicker.filter.supportedDeviceClasses.push(selectedItem.deviceClassSelector);
        }
        else {
            devicePicker.filter.supportedDeviceSelectors.push(selectedItem.selector);
        }

        WinJS.log && WinJS.log("Showing Device Picker...", "sample", "status");

        if (pickSingle) {
            buttonRect = document.getElementById("pickSingleDeviceButton").getBoundingClientRect();
        }
        else {
            buttonRect = document.getElementById("showDevicePickerButton").getBoundingClientRect();
        }

        var rect = { x: buttonRect.left, y: buttonRect.top, width: buttonRect.width, height: buttonRect.height };

        if (pickSingle) {
            devicePicker.pickSingleDeviceAsync(rect).done(function (deviceInfo) {
                if (null != deviceInfo) {
                    resultCollection.push(new DisplayHelpers.DeviceInformationDisplay(deviceInfo));
                }
                showDevicePickerButton.disabled = false;
            });
        }
        else {
            devicePicker.addEventListener("devicepickerdismissed", onPickerDismissed);
            devicePicker.addEventListener("deviceselected", onDeviceSelected);

            // Show the picker
            devicePicker.show(rect);
        }        
    }

    function onPickerDismissed() {
        showDevicePickerButton.disabled = false;
        WinJS.log && WinJS.log("Hiding Device Picker", "sample", "status");
    }

    function onDeviceSelected(args) {
        resultCollection.splice(0, resultCollection.length);
        resultCollection.push(new DisplayHelpers.DeviceInformationDisplay(args.selectedDevice));
    }

})();

File: Samples/DeviceEnumerationAndPairing/cppwinrt/Scenario1_DevicePicker.cpp
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the MIT License (MIT).
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
//*********************************************************

#include "pch.h"
#include "Scenario1_DevicePicker.h"
#include "SampleConfiguration.h"
#include "DeviceHelpers.h"
#include "Scenario1_DevicePicker.g.cpp"

using namespace winrt;
using namespace winrt::Windows::Devices::Enumeration;
using namespace winrt::Windows::Foundation;
using namespace winrt::Windows::UI::Xaml;
using namespace winrt::Windows::UI::Xaml::Media;
using namespace winrt::Windows::UI::Xaml::Navigation;

namespace winrt::SDKTemplate::implementation
{
    Scenario1_DevicePicker::Scenario1_DevicePicker()
    {
        InitializeComponent();
    }

    void Scenario1_DevicePicker::OnNavigatedTo(NavigationEventArgs const&)
    {
        resultsListView().ItemsSource(resultCollection);

        selectorComboBox().ItemsSource(DeviceSelectorChoices::DevicePickerSelectors());
        selectorComboBox().SelectedIndex(0);
    }

    void Scenario1_DevicePicker::OnNavigatedFrom(NavigationEventArgs const&)
    {
        devicePicker = nullptr;
    }

    void Scenario1_DevicePicker::PickSingleDeviceButton_Click(IInspectable const&, RoutedEventArgs const&)
    {
        ShowDevicePicker(
            true // pickSingle
        );
    }

    void Scenario1_DevicePicker::ShowDevicePickerButton_Click(IInspectable const&, RoutedEventArgs const&)
    {
        ShowDevicePicker(
            false // pickSingle
        );
    }

    fire_and_forget Scenario1_DevicePicker::ShowDevicePicker(bool pickSingle)
    {
        auto lifetime = get_strong();

        showDevicePickerButton().IsEnabled(false);
        resultCollection.Clear();

        devicePicker = DevicePicker();

        // First get the device selector chosen by the UI.
        IInspectable item = selectorComboBox().SelectedItem();
        DeviceSelectorInfo* deviceSelectorInfo = FromInspectable<DeviceSelectorInfo>(item);

        if (deviceSelectorInfo->Selector().empty())
        {
            // If the a pre-canned device class selector was chosen, call the DeviceClass overload
            devicePicker.Filter().SupportedDeviceClasses().Append(deviceSelectorInfo->DeviceClassSelector());
        }
        else
        {
            // Use AQS string selector from dynamic call to a device api's GetDeviceSelector call
            devicePicker.Filter().SupportedDeviceSelectors().Append(deviceSelectorInfo->Selector());
        }

        rootPage.NotifyUser(L"Showing Device Picker", NotifyType::StatusMessage);

        // Calculate the position to show the picker (right below the buttons)
        GeneralTransform ge = pickSingleDeviceButton().TransformToVisual(nullptr);
        Point point = ge.TransformPoint({ 0, 0 });
        Rect rect{ point,
                        { static_cast<float>(pickSingleDeviceButton().ActualWidth()),
                          static_cast<float>(pickSingleDeviceButton().ActualHeight()) }
        };

        if (pickSingle)
        {
            DeviceInformation di = co_await devicePicker.PickSingleDeviceAsync(rect);
            if (nullptr != di)
            {
                resultCollection.Append(make<DeviceInformationDisplay>(di));
            }
            showDevicePickerButton().IsEnabled(true);
        }
        else
        {
            devicePicker.DevicePickerDismissed({ get_weak(), &Scenario1_DevicePicker::OnDevicePickerDismissed });
            devicePicker.DeviceSelected({ get_weak(), &Scenario1_DevicePicker::OnDeviceSelected });

            // Show the picker
            devicePicker.Show(rect);
        }
    }

    fire_and_forget Scenario1_DevicePicker::OnDevicePickerDismissed(DevicePicker const&, IInspectable const&)
    {
        auto lifetime = get_strong();

        // Since we have the collection databound to a UI element, we need to update the collection on the UI thread.
        co_await resume_foreground(Dispatcher());

        showDevicePickerButton().IsEnabled(true);
        rootPage.NotifyUser(L"Hiding Device Picker", NotifyType::StatusMessage);
    }

    fire_and_forget Scenario1_DevicePicker::OnDeviceSelected(DevicePicker const&, DeviceSelectedEventArgs args)
    {
        auto lifetime = get_strong();

        // Since we have the collection databound to a UI element, we need to update the collection on the UI thread.
        co_await resume_foreground(Dispatcher());

        resultCollection.Clear();
        resultCollection.Append(make<DeviceInformationDisplay>(args.SelectedDevice()));
    }
}
